src/os/bluestore/BitmapFreelistManager.cc

```diff
@@ -0,0 +1,533 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "BitmapFreelistManager.h"
+#include "kv/KeyValueDB.h"
+#include "kv.h"
+
+#include "common/debug.h"
+
+#define dout_subsys ceph_subsys_bluestore
+#undef dout_prefix
+#define dout_prefix *_dout << "freelist "
+
+void make_offset_key(uint64_t offset, std::string *key)
+{
+  key->reserve(10);
+  _key_encode_u64(offset, key);
+}
+
+struct XorMergeOperator : public KeyValueDB::MergeOperator {
+  virtual void merge_nonexistant(
+    const char *rdata, size_t rlen, std::string *new_value) override {
+    *new_value = std::string(rdata, rlen);
+  }
+  virtual void merge(
+    const char *ldata, size_t llen,
+    const char *rdata, size_t rlen,
+    std::string *new_value) {
+    assert(llen == rlen);
+    *new_value = std::string(ldata, llen);
+    for (size_t i = 0; i < rlen; ++i) {
+      (*new_value)[i] ^= rdata[i];
```

Comment:
Hi Sage, sorry for the late. One question here about merge.
(1), alloc the first block ===> txn->merge(bitmap_prefix, k, "10000000");
(2), release the first block ===> txn->merge(bitmap_prefix, k, "00000000");
Then we would merge the them. from what I read from merge() here, it should be:
```"10000000" ^ "00000000" = "10000000"``` 
But,,,,it should "00000000" after a release I think. There must be something wrong in my understanding, please correct me, thanx.